### ISSUE TITLE
loss=nan

### ISSUE THREAD

**Issue by yyhlvdl** [taolei87/sru#53]
@taolei87 hi, I want to thank you very much firstly. Your attribution about SRU helps me a lot. Seeing models with SRU running so fast makes me live in dream. So, I deploy SRU to my own code, replacing lstm(gru). However, my new model with SRU always has the problem: loss=nan. I study previous issues and found that we should use gradient clipping. Can you give some example about how to set gradient clipping, because I did not find in codes published? And the new 'Pull requests' may talk about it.Wish more and more people enjoy the speed of SRU, it's magic!

**Comment by taoleicn** [comment]
Hi @yyhlvdl 

Did you use layer norm? Are you working on a NLP problem?
Did you do some debugging to check at which place the NaN first happen? 
Did lowering the learning rate help?

It is hard to tell without some context information about your problem / code / model..

gradient clipping is used in the language modeling example:
https://github.com/taolei87/sru/blob/master/language_model/train_lm.py#L140

**Comment by yyhlvdl** [comment]
@taoleicn  Thank your answer. I didn't use layer norm, and I try to employ SRU to image. In fact, loss=nan is happened sometimes, i.e., this time, loss=nan; next time ,loss is normal. I think may be I use softmax rather than logsoftmax, however, other nets without SRU won't happen loss=nan. My part code like this:
class DBSRU(nn.Module):
    def __init__(self, src_size,hidden_size):
        super(DBSRU, self).__init__()                                        
        self.hidden_size = hidden_size                                   #(H)
        self.sru=SRU(src_size, hidden_size,
                num_layers = 1,          # number of stacking RNN layers
                dropout = 0.0,           # dropout applied between RNN layers
                rnn_dropout = 0.0,       # variational dropout applied on linear transformation
                use_tanh = 1,            # use tanh?
                use_relu = 0,            # use ReLU?
                use_selu = 0,            # use SeLU?
                bidirectional = True,   # bidirectional RNN ?
                weight_norm = False,     # apply weight normalization on parameters
                layer_norm = False,      # apply layer normalization on the output of each layer
                highway_bias = 0         # initial bias of highway gate (<= 0)
)
    def forward(self,src):
        #BSRU编码:(T,B,W) -> (T,B,number of directions*H);(layers,B, number of directions * H)
        encoder_outputs,hidden = self.sru(src)       
        #(T,B,2H) -> (T,B,H)
        encoder_outputs = (encoder_outputs[:, :, :self.hidden_size] +
                   encoder_outputs[:, :, self.hidden_size:])

       #(layers,B, number of directions * H)-> (layers,B,H)
        #hidden = (hidden[:, :, :self.hidden_size] +
         #          hidden[:, :, self.hidden_size:])

        #(T,B,H) -> (H,B,T)
        encoder_outputs = encoder_outputs.transpose(0,2)
        #(H,B,T) -> (H,B,2H)-> (H,B,H)
        encoder_outputs,hidden = self.sru(encoder_outputs)       
        encoder_outputs = (encoder_outputs[:, :, :self.hidden_size] +
                   encoder_outputs[:, :, self.hidden_size:])
        
        return encoder_outputs
I use it to reduce image size, w-w/2, without MAXPOOL. And I also use SRU to act as decoder:
 output, hidden = self.sru(rnn_input)
